### PR TITLE
Week05

### PR DESCRIPTION
**Pull request by wolves**
Summary
---------------------------------------------------
*Summarize what is included in this PR?*

This Pull Request includes a full test suite to provide 100% test coverage for the existing code found in the `clause.go`, `errors.go`, and `store.go` files.

Code Description
------------------------------------------
*Briefly describe what your code intends to do*

### Clause
```Go
type Clauses map[string]interface{}


func (cls Clauses) String() string { ... }


func (cls Clauses) Match(m Model) bool { ... }
```

The `clause.go` file is relatively small and basic in its functionality. Its overarching functionality is based on mapping keys and values that will serve as a way to query the `Store` for any models that match via its `Match` method.

It also provides a `String()` method that currently is utilized within the relevant errors.

### Testing Implementation

The approach for applying tests to this file was pretty straightforward and divided into two parts between the `String` and the `Match` method functionality.

For the `String()` method the two surface points that required testing were ensuring proper formatting of its returned `string` value and the sorting that it applies to the `Clauses` map. This was accomplished by applying  a test table that exercises this method against the following cases:

```Go
emptyClauses := Clauses{}
mixedClauses := Clauses{"a": 1, "b": "two"}
unorderedClauses := Clauses{"zed": 1, "alpha": 10, "alph4": 11}
complexClauses := Clauses{"1": st, "2": ss}

// Here `st` is of type:
// struct {
//    ordinal int
//    desc    string
// }
// and `ss` []string
```

For the `Match()` method a I created a simple sample of test `Clauses`:

```Go
c := Clauses{"a": 1, "b": 2}
```

I then constructed another test table and provided it a set of varying `Model`'s to run through those `Clauses.Match()` method that looked like this:

```Go
matchModel := Model{"a": 1, "b": 2}
noMatchModel := Model{"nope": "z"}
emptyModel := Model{}
```

### Errors

`errors.go` is based around the two custom errors `ErrTableNotFound` and `ErrNoRows` and their corresponding methods. These serve as error handling for the datastore when a non-existent table query or non-matching clause query are made respectively. 

### Testing Implementation
Being that the errors have some testing surface exposed within the `Store` that ultimately gets tested through its integration my intentions here were to directly check the behavior of each errors methods and functions against not only the happy path and the main failure path but also as many edge cases as I believed were feasible possibilities. This was all accomplished with table tests and verification of the resulting outputs for each case.

With regard to `ErrTableNotFound`, it was tested against the following 2 test cases:
```Go
errWithTable := ErrTableNotFound{tn}
errWithoutTable := ErrTableNotFound{}
```

The `errors.Is` functions for this error type were then verified along with its error message formatting and the `TableNotFound()` methods output.

For the `ErrNoRows` error type there were a few more scenarios to cover. It provides a couple of methods for accessing the table name and `Clauses` associated with it as well as functions for `errors.Is` and `errors.As` implementations and the standard error message value that includes the aforementioned table name and `Clauses` that triggered it.

For the pointer methods associated with the `ErrNoRows` query data and error message I used these table tests:
```Go
{name: "with values", err: errWithValues, exp_table: tn, exp_clauses: c},
{name: "without values", err: errEmpty, exp_table: "", exp_clauses: Clauses{}},
{name: "without table", err: errWithoutTable, exp_table: "", exp_clauses: Clauses{}},
{name: "without clauses", err: errWithoutClauses, exp_table: tn, exp_clauses: Clauses{}},
{name: "with nil clauses", err: errWithNilClauses, exp_table: tn, exp_clauses: Clauses{}},
```

The resulting error messages and tracked data values were then verified.

For the functionality that interfaces with `errors.Is` and `errors.As` the following test scenarios were used:
```Go
{name: "with noRowError", err: errWithValues, exp: true},
{name: "with nil error", err: nil, exp: false},
{name: "with ErrTableNotFound", err: ErrTableNotFound{}, exp: false},
```

### Store

`store.go` provides a basic data store along with methods to insert, select, list, or that assist in these actions. It is also the place where the functionality from the other two files coalesces.

### Testing Implementation

The `Insert()` and `Len()` functionality along with any resulting values and/or errors were tested with the following test cases:
```Go
{ name: "with empty store", store: emptyStore, query: tn, exp_len: 0, exp_err: ErrTableNotFound{table: tn} },
{ name: "with empty data", store: withEmptyData, query: tn, exp_len: 0, exp_err: ErrTableNotFound{table: tn} },
{ name: "with empty models", store: withEmptyTable, query: tn, exp_len: 0, exp_err: nil },
{ name: "with empty store", store: withPopulatedStore, query: tn, exp_len: 1, exp_err: nil },
```

The methods `All()`, `Select()` and the private integrated `db()` function were exercised via their coupled functionality using these test scenarios:
```Go
{ name: "with table and query match", store: withPopulatedStore, table_name: tn, query: q, exp_mods: Models{mod}, exp_err: nil },
{ name: "with no table", store: emptyStore, table_name: tn, query: q, exp_mods: nil, exp_err: ErrTableNotFound{table: tn} },
{ name: "with empty query", store: withPopulatedStore, table_name: tn, query: Clauses{}, exp_mods: Models{mod}, exp_err: nil },
{ name: "with no query matches", store: withPopulatedStore, table_name: tn, query: Clauses{"no": "match"}, exp_mods: nil, exp_err: &errNoRows{clauses: Clauses{"no": "match"}, table: tn} },

```

The expectations of the data and errors related to those cases were tested and verified and with all of that, I achieved coverage that wasn't just full but also that I fell is fairly tidy. I also believe the tests are not simply testing for the sake of coverage but testing the actual logic and surface area to verify edges cases and all functionality.

Essay/Commentary
------------------------------------------
*A long-form writeup covering: Difficulties encountered / Noteworthy choices made / Other things that stood out to you*

No pomp and circumstance let's cut directly to it. My submission for this week was the result of a battle, a proverbial bar brawl...what felt like a knockdown, drag-out, no-holds-barred assault that tested my willpower. I can honestly say that with my current level of experience I gave it just about everything I have. Looking at it now in its successfully finished state I can say that it not only meets the criteria set forth but also it does literally the best I am currently capable of. I say this now, because I know I will most likely later come back and have enough knowledge to accomplish this with much less effort. I digress though. Let me briefly elaborate.

I am pretty familiar with testing and enjoy using it often so things started relatively smooth. To be honest a large majority of the implementation went relatively swiftly. I then hit the infamous last little "20%" of the process. I was able to get things to 100% coverage but not in a way that I would feel good about submitting. So I refactored things. I threw away whole pieces and completely rewrote them in ways that I felt were better and would more accurately test the actual code not just cheese their way to full coverage. Then finally at some point, I got stuck. I had somewhere between 94 and 98 percent coverage but was bouncing back and forth between test panics caused by a nil error check test on the `ErrNoRows` `errors.As` implementation methods or adjusting things to correct for it that would in-turn result in missing coverage or code that I knew was not right. I was not willing to give up and just throw in the code I knew smelled but gave me 100% coverage. So I beat my head against the problem for longer than I am willing to admit. We'll just say I got to see the sun travel in both directions without leaving my editor. This is not at all a result of the complexity or difficulty of the assignment but just one of those moments in learning when you get stuck and it requires sheer determination to get through it. It will be a moment I will always remember and the knowledge I gained through it will stick with me. I hope the result is satisfactory and I look forward to any other suggested improvements that I can learn from and add.

### CODE REVIEW

File: week05/clause_test.go [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+import (
+	"fmt"
+	"testing"
+)
+
+func Test_Clause(t *testing.T) {
+	t.Parallel()
+
+	t.Run("String method", func(t *testing.T) {
+		t.Parallel()
```
Comment (markbates) [inline week05/clause_test.go:12]: As we talked about in class, you should removed the `t.Parallel()` calls to make sure your subtests complete.

File: week05/store_test.go [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+	withPopulatedStore.Insert(tn, mod)
+
+	t.Run("Len", func(t *testing.T) {
+		t.Parallel()
+
+		table := []struct {
+			name    string
+			store   *Store
+			query   string
+			exp_len int
```
Comment (markbates) [inline week05/store_test.go:30]: In Go we use Camel/PascalCase for naming. 

```suggestion
			expLen int
```

**Review by markbates — COMMENTED** [review]
Excellent work.
